src/core/SampleDecoder.cpp

```diff
@@ -0,0 +1,167 @@
+#include "SampleDecoder.h"
+
+#include <QFile>
+#include <QFileInfo>
+#include <QString>
+#include <memory>
+#include <sndfile.h>
+
+#ifdef LMMS_HAVE_OGGVORBIS
+#include <vorbis/vorbisfile.h>
+#endif
+
+#include "AudioEngine.h"
+#include "DrumSynth.h"
+#include "Engine.h"
+#include "lmms_basics.h"
+
+namespace lmms {
+
+std::vector<SampleDecoder::Decoder> SampleDecoder::s_decoders = {
+	&SampleDecoder::decodeSampleSF,
+#ifdef LMMS_HAVE_OGGVORBIS
+	&SampleDecoder::decodeSampleOggVorbis,
+#endif
+	&SampleDecoder::decodeSampleDS
+};
+
+auto SampleDecoder::supportedAudioTypes() -> std::vector<AudioType>
+{
+	auto sfFormatInfo = SF_FORMAT_INFO{};
+
+	// Add DrumSynth by default since that support comes from us
+	auto audioTypes = std::vector<AudioType>{AudioType{"DrumSynth", "ds"}};
+
```

Comment:
```suggestion
auto SampleDecoder::supportedAudioTypes() -> const std::vector<AudioType>&
{
	static auto audioTypes = std::vector<AudioType>{};
	if (!audioTypes.empty()) { return audioTypes; }

	auto sfFormatInfo = SF_FORMAT_INFO{};

	// Add DrumSynth by default since that support comes from us
	audioTypes.push_back(AudioType{"DrumSynth", "ds"});

```
This method only needs to get the supported audio types from sndfile once, and can cache the results for future calls.

Reply:
I don't think this optimal. The creation of `audioTypes` is only made once, but the overall function still gets ran. That would also mean the function would work on the same vector for multiple calls, which is a problem. Could use `std::call_once`, but storing a `static std::vector<AudioTypes>` is easier.